src/concurrency/thread.rs

```diff
@@ -620,11 +630,26 @@ impl<'mir, 'tcx: 'mir> ThreadManager<'mir, 'tcx> {
     /// long as we can and switch only when we have to (the active thread was
     /// blocked, terminated, or has explicitly asked to be preempted).
     fn schedule(&mut self, clock: &Clock) -> InterpResult<'tcx, SchedulingAction> {
-        // Check whether the thread has **just** terminated (`check_terminated`
-        // checks whether the thread has popped all its stack and if yes, sets
-        // the thread state to terminated).
-        if self.threads[self.active_thread].check_terminated() {
-            return Ok(SchedulingAction::ExecuteDtors);
+        // If we are the main thread, and our call stack is empty but our state is Enabled, we are
+        // about to terminate.
+        // But, if there are any other threads which can execute, yield to them instead of falling
+        // through to the termination state.
+        let active_thread = &self.threads[self.active_thread];
+        if self.active_thread == MAIN_THREAD
+            && active_thread.stack.is_empty()
+            && active_thread.state == ThreadState::Enabled
```

Comment:
This is the same check as in `check_terminated`, right? I think we should make that function really just *check* and not also change the state -- this has confused be in the past several times. Then we can also reuse it here, right?